cmake: kconfig: introduce dedicated unit testing board and support Kconfig for unit tests

tejlmand opened the pull request:
This PR introduces a dedicated unit testing board.

Today, a dedicated Zephyr unit testing scheme exists but is different
from how a Zephyr build generally works.

For example Kconfig is not possible, resulting on various different
hacks to pass Kconfig settings from test cases / testcase.yaml through
CMake to the code.

Some directly as compile definitions, some as header files with forced
inclusion on sources, some with wrapper flags which again results in
different define being enabled. There is even cases where a second
forced header inclusion undefines previous defines.

Unit test often does a manual check for the right boards, like this:
```
if (NOT BOARD STREQUAL unit_testing)
    message(FATAL_ERROR "This project can only be used with...")
endif()
```

Introducing a dedicated unit_testing board under `tests/root` allows
us to use Kconfig in unit test samples, and thus proper `prj.conf` and
extra Kconfig fragments.
Generation of autoconf.h so the overall architecture follows regular
Zephyr builds.

Proper and uniform error messages when invalid board is selected.

The unit_testing board and arch is located under: `tests/root` so that
it is only available when `find_package(Zephyr COMPONENTS unittest)`
is used, and not available for regular Zephyr builds.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
scripts: remove ZEPHYR_BASE as default root for board listing

Remove ZEPHYR_BASE as default root in list_boards.py.

This allows list_boards.py to be used to only print boards at given
root(s) without printing Zephyr default boards.

Secondly it remove the need in list_boards.py to have any knowledge of
ZEPHYR_BASE.

The `west boards` command already has ZEPHYR_BASE knowledge and can
easily add ZEPHYR_BASE to the list of roots it is already constructing,
thus removing the need for knowing that ZEPHYR_BASE is always added.

Update boards.cmake to pass ZEPHYR_BASE as an additional arch root, as
arch.cmake has not yet been processed,which means ZEPHYR_BASE is missing
in ARCH_ROOT list at this point in time.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: define and create KCONFIG_BINARY_DIR if it doesn't exists

The Zephyr CMake module kconfig.cmake will create files inside
KCONFIG_BINARY_DIR.

In a default Zephyr build setup, this directory is created by Zephyr
CMake module zephyr_module.cmake. But Zephyr CMake modules are modular
and thus it must be possible to use kconfig module without
zephyr_module.

Thus kconfig.cmake must set and create KCONFIG_BINARY_DIR if it does not
already exists.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
kconfig: optional sourcing of $(KCONFIG_BINARY_DIR)/Kconfig.modules

Change sourcing of generated Kconfig.module to be optional.

It should be possible to run parse the Zephyr Kconfig tree even if the
Kconfig.modules file has not been generated.

This can be useful in testing where you may wish to run without Zephyr
modules present, and thus not load the Zephyr module CMake module.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
dts: optional sourcing of $(KCONFIG_BINARY_DIR)/Kconfig.dts

Change sourcing of generated Kconfig.dts to be optional.

It should be possible to run parse the Zephyr Kconfig tree even if the
Kconfig.dts file has not been generated.

This can be useful in unit testing where DTS may not be executed.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
kconfig: change $(ARCH_DIR) to arch/

Changing $(ARCH_DIR)/common/Kconfig to arch/common/Kconfig.

The use of ARCH_DIR at this place is wrong, as it suddenly requires out
of tree archs to support a common/Kconfig file, which may make no sense
to them.

If an out of tree arch wants to place common Kconfig code in a common
Kconfig file, that's their choice and they should source such file
themselves.

Instead just source the Zephyr arch common file directly.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: kconfig: introduce dedicated unit testing board

This commit introduces a dedicated unit testing board.

Today, a dedicated Zephyr unit testing scheme exists but is different
from how a Zephyr build generally works.

For example Kconfig is not possible, resulting on various different
hacks to pass Kconfig settings from test cases / testcase.yaml through
CMake to the code.
Some directly as compile definitions, some as header files with forced
inclusion on sources, some with wrapper flags which again results in
different define being enabled. There is even cases where a second
forced header inclusion undefines previous defines.

Unit test often does a manual check for the right boards, like this:
> if (NOT BOARD STREQUAL unit_testing)
>    message(FATAL_ERROR "This project can only be used with...")
> endif()

Introducing a dedicated unit_testing board under `tests/root` allows
us to use Kconfig in unit test samples, and thus proper `prj.conf` and
extra Kconfig fragments.
Generation of autoconf.h so the overall architecture follows regular
Zephyr builds.

Proper and uniform error messages when invalid board is selected.

The unit_testing board and arch is located under: `subsys/testsuite` so
that it is only available when find_package(Zephyr COMPONENTS unittest)
is used, and not available for regular Zephyr builds.

Kconfig generates autoconf.h which is applied as compile flag to
test binary which means that kconfig defines placed in ztest.h can now
be removed.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
tests: remove testing for board unit_testing

Remove boilerplate code:
> if (NOT BOARD STREQUAL unit_testing)
>    message(FATAL_ERROR "This project can only be used with...")
> endif()

as that is now handled by boards.cmake when validating the board.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
@ahmedmoheb-nordic cannot add you as reviewer, but please test this PR and give comments.

tejlmand commented:
@ppryga and @thoh-ot this should allow you to cleanup what was in the thread here: https://github.com/zephyrproject-rtos/zephyr/pull/40073#issuecomment-1041451330

Would be great if you can try it out just on a few existing test cases that was merged in #40073 so we can know if more is needed in this PR.

ppryga commented:
@tejlmand I'll try to find some time to look at this PR and do some tests next week.

thoh-ot commented:
@tejlmand Very awesome, I will give it a spin next week

tejlmand committed:
tests: adopt ztest base test to extended unittest module

With the support of Kconfig in unittest CMake module then prj.conf is
now supported.

Add a prj.conf with ZTEST config settings and remove hardcoded values
from CMakeLists.txt file.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
tests: add Kconfig file to bluetooth ctrl_data_length_update testcases

With Kconfig support in unittest the bluetooth ctrl_data_length_update
test fails because the symbols being configured are promptless or
disabled because of dependencies.

Add a Kconfig file to make needed symbols visible in unittest mode.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented on tests/bluetooth/controller/ctrl_data_length_update/Kconfig:
```diff
@@ -0,0 +1,19 @@
+#
+# Copyright (c) 2022, Nordic Semiconductor ASA
+
+config BT_CTLR_PHY
+	prompt "Bluetooth 5.0 phy"
+	help
+	  This is generally a hidden symbol but for unit testing a prompt is
+	  defined to allow controlling this setting in the build.
```
@thoh-ot the fix here in this commit 4e90b63af06bc936b4640cc8a5628c2dab9a753b is not necessarily the cleanest approach.

I believe the proper approach would be to create a `prj.conf` with needed settings similar to this commit: https://github.com/zephyrproject-rtos/zephyr/pull/48732/commits/e58fab9d739611045d9b8b5d6ce9df62a15d0a73
but the `BT_CTLR_PHY` has indirect dependencies to many settings, including `BT_LLL_VENDOR_NORDIC` or `BT_LLL_VENDOR_OPENISA` so unit testing would have to specify one of those symbols.

So in the end, the cleanest solution for unit testing might actually be to bypass dependencies for settings as is done in this commit.

thoh-ot replied:
@tejlmand Been working with the PR today, but havn't gotten a conclusion yet. There is some trickery in the `tests/bluetooth/controller/common/defaults_cmake.txt` that i'm working on removing and that gives me a tons of compile errors due to all this missing Kconfig symbols, so i'm trying to work around that.

yperess commented:
@tejlmand this is great, I was 1/2 way doimg the exact same thing. Thank you!!!!

jeremybettis reviewed (commented):
In https://github.com/zephyrproject-rtos/zephyr/pull/48732/commits/c373900fc18238d64288d561068c638ce084224c#:~:text=Kconfig%20file%2C%20that%27s-,there,-choice%20and%20they you meant `their` not `there`.

yperess commented on arch/Kconfig:
```diff
@@ -149,7 +149,7 @@ config ARCH_IS_SET
 menu "General Architecture Options"
 
-source "$(ARCH_DIR)/common/Kconfig"
+source "arch/common/Kconfig"
```
I think you need to keep this for downstream users that want to override it no? In my local change I set `${ARCH_DIR}`  in unittest.cmake to `${ZEPHYR_BASE}/arch` if it wasn't already set

tejlmand replied:
> I think you need to keep this for downstream users that want to override it no? 

The `source "$(ARCH_DIR)/common/Kconfig"` was introduced here #44358 in commit 070422db46af53cd692f01320b410d3ba07e57df but should never have been introduced in the first place.

This change foces every user with an out-of-tree arch to implement a `common/Kconfig` file, even if they don't need it.
If users wants to create a common Kconfig for all their arch and source such file, they can do that themselves, Zephyr shouldn't require such file to be present.

>  In my local change I set `${ARCH_DIR}` in unittest.cmake to `${ZEPHYR_BASE}/arch` if it wasn't already set

and you can still do that, and your arch Kconfig will be picked up from here:
https://github.com/zephyrproject-rtos/zephyr/blob/ccc0315e14341879cc466156a4c925dd27b06525/arch/Kconfig#L12
for your custom arch.

yperess commented on cmake/modules/root.cmake:
```diff
@@ -32,3 +32,10 @@ zephyr_file(APPLICATION_ROOT SOC_ROOT)
 # Convert paths to absolute, relative from APPLICATION_SOURCE_DIR
 zephyr_file(APPLICATION_ROOT ARCH_ROOT)
+
+if(unittest IN_LIST Zephyr_FIND_COMPONENTS)
+  # Zephyr used in unittest mode, use dedicated unittest root.
+  set(BOARD_ROOT ${ZEPHYR_BASE}/tests/root)
+  set(ARCH_ROOT  ${ZEPHYR_BASE}/tests/root)
+  set(SOC_ROOT   ${ZEPHYR_BASE}/tests/root)
+endif()
```
Maybe set these in unittest.cmake?

tejlmand replied:
no, the introduction of Zephyr CMake modules and its modularization here https://github.com/zephyrproject-rtos/zephyr/pull/41301 tries to improve the overall design and reuse of code.

This file deals with root handling, that is used by other modules.
If you set those in unittest.cmake you must set those BEFORE any `include()`, but that breaks the design as you loose some features of the modular design.
All Zephyr CMake modules (except unittest and zephyr_default that a special cases) are having include guards:
https://github.com/zephyrproject-rtos/zephyr/blob/ccc0315e14341879cc466156a4c925dd27b06525/cmake/modules/root.cmake#L20

This allows code to include them multiple times and only have them executed the first time.
This again means that users can do neat trick like this, before loading, for example unittest.

Normal:
```
find_package(Zephyr COMPONENTS unittest)
```

Locally adjusted root result:
```
list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/modules_tweaks)
find_package(Zephyr COMPONENTS root root_tweak unittest)
```

and then inside `root_tweak` decide what should happen to any root.
Then when `unittest` is loaded, the result from `root_tweak` is used.
Such possibility is lost if we go back to old style boilerplate of placing everything in just one file.

The cleanup and design work done in #41301 is exactly what make it possible to easily reuse Kconfig know.
Although there are still a lot of improvement to be done.

For a similar example, look here:
https://github.com/zephyrproject-rtos/zephyr/blob/ccc0315e14341879cc466156a4c925dd27b06525/share/sysbuild/CMakeLists.txt#L16-L23

yperess replied:
That's really neat, thanks for the detailed explanation, looks like I have some more to learn about our build system :)

yperess commented on tests/root/arch/unit_testing/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (c) 2022 Nordic Semiconductor
+
+# Intentionally left empty
```
Why are all these under `tests/` instead of at the top level `arch/unit_testing/`? I wouldn't have found it if I wasn't reviewing this PR

tejlmand replied:
because if they are placed under toplevel `arch/unit_testing` then that arch is always available, and I really don't like that.

The `unit_testing` board and arch are not really board and arch in the normal sense, instead they are solving a very special testing use-case.

But i'm open to another location than `tests/root/<board|arch>` cause i'm not particularly fond of that location either.
Just lack a better candidate, as `unit_testing` should not pollute `<ZEPHYR_BASE>/boards` and `<ZEPHYR_BASE>/arch`

yperess replied:
Can we put it under `subsys/testsuite/arch/`?

tejlmand replied:
Not a bad location.

So then we would have:
`subsys/testsuite/arch/`
`subsys/testsuite/boards/`

for what is related to unittesting.

Agreed ?

yperess replied:
Yup LGTM

tejlmand replied:
done.

yperess replied:
Thanks!

yperess reviewed (changes requested):
Thank you so much for this. I believe that it would be easier to follow if the arch/unit_testing was added at the top level. Also, the various configuration values should just be in unittest.cmake to keep it all in one place, if we ever need to make further changes we don't want to have to hunt things down.

tejlmand commented:
> In [c373900#:~:text=Kconfig%20file%2C%20that%27s-,there,-choice%20and%20they](https://github.com/zephyrproject-rtos/zephyr/commit/c373900fc18238d64288d561068c638ce084224c#:~:text=Kconfig%20file%2C%20that%27s-,there,-choice%20and%20they) you meant `their` not `there`.

@jeremybettis  I sure did.
The good old classic of you fingers typing something else than your thinking.
Fixed, and now your +1 is gone 😞

ahmedmoheb-nordic commented:
> @ahmedmoheb-nordic cannot add you as reviewer, but please test this PR and give comments.

@tejlmand I think there is a problem that not all defined configuration parameters were added as a compilation defines.

I added the following configurations to a unit testing project which should make `CONFIG_BT_HCI_ACL_FLOW_CONTROL` defined in source files

`CONFIG_BT=y
CONFIG_BT_HCI=y
CONFIG_BT_PERIPHERAL=y
CONFIG_BT_CENTRAL=y
CONFIG_BT_HCI_ACL_FLOW_CONTROL=y`

when I added the following code to one of the test function, it was compiled and passed without reporting any issues
`#ifdef CONFIG_BT_HCI_ACL_FLOW_CONTROL
#error "This isn't allowed CONFIG_BT_HCI_ACL_FLOW_CONTROL"
#endif`

I tested the same approach with the twister integration project at `/samples/subsys/testsuite/integration` and the compilation process was teminated due to hitting `#error`

tejlmand commented:
> @tejlmand I think there is a problem that not all defined configuration parameters were added as a compilation defines.

Seems `autoconf.h` is not applied on sources being compiled.
Looking into this.

`DNM` for now.

tejlmand commented:
> Seems `autoconf.h` is not applied on sources being compiled. Looking into this.
> 
> `DNM` for now.

fixed

yperess commented:
@tejlmand seems there are a few things missing from linking.

tejlmand commented:
> @tejlmand seems there are a few things missing from linking.

yes, it seems that the autoconf.h is impacting test when building for 64 bit.
Not had time to investigate fully.

tejlmand committed:
tests: update unit/cbprintf test to use Kconfig

This commit removes CONFIG defines from main.c and instead uses the
proper Kconfig handling of specifying extra Kconfig options in
testcase.yaml.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
> > @tejlmand seems there are a few things missing from linking.
> 
> yes, it seems that the autoconf.h is impacting test when building for 64 bit. Not had time to investigate fully.

it was actually the assert, just got tricked because the tests I initially looked where this happened was 64bit builds of cbprintf, not it's 32 bit counterpart.

Should be fixed now.

yperess commented:
> > > @tejlmand seems there are a few things missing from linking.
> > 
> > 
> > yes, it seems that the autoconf.h is impacting test when building for 64 bit. Not had time to investigate fully.
> 
> it was actually the assert, just got tricked because the tests I initially looked where this happened was 64bit builds of cbprintf, not it's 32 bit counterpart.
> 
> Should be fixed now.

Can we instead set the default to assert if the board is not the unit_test board?

tejlmand commented:
> Can we instead set the default to assert if the board is not the unit_test board?

I was starting to think the same, so that is now done.

dcpleung commented:
Wonder if `subsys/testsuite/{arch,board,soc}` should be moved down one level. It looks weird if `arch`, `board` and `soc` appear alongside `coverage` and `ztest`.

tejlmand commented:
> Wonder if `subsys/testsuite/{arch,board,soc}` should be moved down one level. It looks weird if `arch`, `board` and `soc` appear alongside `coverage` and `ztest`.

Good question, although i'm unsure what the name should be.
Afterall, the real Zephyr folders `arch|soc|boards` are at the top-level of Zephyr repo.

I used `/tests/root/[boards|arch|soc]` in first version, but never really liked it (especially because the `root` part), so was happy with the `/subsys/testsuite` proposal.

ahmedmoheb-nordic commented:
> > Seems `autoconf.h` is not applied on sources being compiled. Looking into this.
> > `DNM` for now.
> 
> fixed

I tested it and it's working now

dcpleung commented:
> > Wonder if `subsys/testsuite/{arch,board,soc}` should be moved down one level. It looks weird if `arch`, `board` and `soc` appear alongside `coverage` and `ztest`.
> 
> Good question, although i'm unsure what the name should be. Afterall, the real Zephyr folders `arch|soc|boards` are at the top-level of Zephyr repo.
> 
> I used `/tests/root/[boards|arch|soc]` in first version, but never really liked it (especially because the `root` part), so was happy with the `/subsys/testsuite` proposal.

Maybe `subsys/testsuite/overlay/{arch,board,soc}`? These configs can be considered overlays on top of the top level arch,board,soc configs.

tejlmand commented:
> Maybe `subsys/testsuite/overlay/{arch,board,soc}`? These configs can be considered overlays on top of the top level arch,board,soc configs.

But they are not overlays, not in any way.
Starting to refer to them as overlays or in such folder will really add to confusion.

This is an independent arch, soc, boards tree.

If I try to build for a wrong board in a regular Zephyr CMake build, I will be printed with a list of valid boards under their respective archs like this:
```
-- Board: invalid_board
No board named 'invalid_board' found.

Please choose one of the following boards:

arc:
  em_starterkit
  em_starterkit_em11d
  em_starterkit_em7d
  em_starterkit_em7d_v22
....
arm:
  96b_aerocore2
  96b_argonkey
  96b_avenger96
...
```
but the board `unit_testing` is not in that list, because it's not a valid board in a normal project.

But it's a valid board in unit testing, so if I do the same on a unit testing project like: https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/unit/crc
I get this:
```
-- Board: invalid_board
No board named 'invalid_board' found.

Please choose one of the following boards:

unit_testing:
  unit_testing
```

and none of the regular boards are available.

That is because it's a completely independent tree of archs and boards.

tejlmand commented:
@yperess and @jeremybettis can I get your +1 back :smile:

ahmedmoheb-nordic commented:
@tejlmand I found another issue related to not including the `autoconf.h` while building sources.

For one of the unit test projects, I tried to build a simple library by adding one of the mocking files `net_buf.c` by adding the followings to one of the projects CMakeLists.txt

`zephyr_library()`
`zephyr_library_compile_definitions(ZTEST_UNITTEST)`
`zephyr_library_include_directories(${CMAKE_BINARY_DIR}/zephyr/include/generated)`
`zephyr_library_include_directories(${ZEPHYR_BASE}/subsys/testsuite/include)`
`zephyr_library_include_directories(${ZEPHYR_BASE}/subsys/testsuite/include/zephyr)`
`zephyr_library_include_directories(${ZEPHYR_BASE}/subsys/testsuite/ztest/include)`
`zephyr_library_include_directories(${ZEPHYR_BASE}/subsys/testsuite/ztest/include/zephyr)`
`zephyr_library_include_directories(${ZEPHYR_BASE}/include)`
`zephyr_library_sources(${ZEPHYR_BASE}/tests/bluetooth/host/buf/mocks/net_buf.c)`

Trying to compile, it gives several errors like

- `error: #error "You need to add CONFIG_ZTEST to your config file."`
- `zephyrproject/zephyr/include/zephyr/toolchain/gcc.h:554:2: error: #error processor architecture not supported`

I think the problem is that the file `autoconf.h` isn't added while compiling the file `net_buf.c`.

tejlmand commented:
> I think the problem is that the file autoconf.h isn't added while compiling the file net_buf.c.

The unit_testing has always been a very special use-case / project, so it has never followed the regular Zephyr CMake approach.
Therefore the `zephyr_interface` is probably not populated.
Also before this PR, the Zephyr extension commands were not available either.

Wonder if we should populate `zephyr_interface` or leave the unit tests more pure, and thus request users creating libraries to append such flags like this:
```
target_compile_options(<library> PRIVATE -imacros ${AUTOCONF_H})
```
or
```
zephyr_library_compile_options(-imacros ${AUTOCONF_H})
```


This is not directly related to Kconfig and unit_testing board (this PR), so would like to do some thinking and create a dedicated PR for a proposal on populating `zephyr_interface` with such flags.

carlescufi commented:
@jeremybettis could you take another look please?

carlescufi commented:
> @jeremybettis could you take another look please?

nvm, I see you already approved earlier.
